src/lambda_handler.py

```diff
@@ -1,22 +1,152 @@
+    role_credentials: Tuple[str, str, str] = assume_role(role_arn, session_name)
+
+    return boto3.resource(
+        aws_service,
+        region_name=region,
+        aws_access_key_id=role_credentials[0],
+        aws_secret_access_key=role_credentials[1],
+        aws_session_token=role_credentials[2],
+    )
+
+
+def convert_tags(aws_resource: boto3.resource) -> Dict[str, str]:
+    """Convert resource tags from an AWS dictionary into a Python dictionary."""
+    try:
+        tags: Dict[str, str] = {x["Key"]: x["Value"] for x in aws_resource.tags}
+    except TypeError:
+        # This happens if there are no tags associated with the resource
+        tags = {}
+    return tags
+
+
+def get_ec2_ips(
+    ec2: boto3.resource, application_tag_name: str, publish_egress_tag_name: str
+) -> Iterator[Tuple[str, str]]:
+    """Create a set of public EC2 IPs.
+
+    Yields (application tag value, public_ip) tuples.
+    """
+    # Get list of running EC2 instances
+    instances: boto3.resources.collection = ec2.instances.filter(
+        Filters=[{"Name": "instance-state-name", "Values": ["running"]}]
+    )
+    # Get list of elastic IPs in all VPCs
+    vpc_addresses: boto3.resources.collection = ec2.vpc_addresses.all()
+
+    for instance in instances:
+        # If the instance doesn't have a public IP, we can skip it
+        if instance.public_ip_address is None:
+            continue
+
+        # Convert instance tags from an AWS dictionary into a Python dictionary
+        tags = convert_tags(instance)
+
+        # If the publish egress tag doesn't exist or isn't set to True, skip it
+        if tags.get(publish_egress_tag_name, str(False)) != str(True):
+            continue
+        # Send back a tuple associating the public IP to an application.
+        # If application is unset, return "", so that the IP can be included
+        # in a list of all IPs if desired (e.g. using app_regex=".*").
+        yield (tags.get(application_tag_name, ""), instance.public_ip_address)
+
+    for vpc_address in vpc_addresses:
+        # Convert elastic IP tags from an AWS dictionary into a Python dictionary
+        eip_tags = convert_tags(vpc_address)
+
+        # If the publish egress tag doesn't exist or isn't set to True, skip it
+        if eip_tags.get(publish_egress_tag_name, str(False)) != str(True):
+            continue
+        # Send back a tuple associating the public IP to an application.
+        # If application is unset, return "", so that the IP can be included
+        # in a list of all IPs if desired (e.g. using app_regex=".*").
+        yield (eip_tags.get(application_tag_name, ""), vpc_address.public_ip)
+
+
+def get_ec2_regions(
+    ec2: boto3.client, filter: Optional[List[Dict[str, Union[str, List[str]]]]] = []
+) -> List[str]:
+    """Get a filtered list of all the regions with EC2 support."""
+    response = ec2.describe_regions(Filters=filter)
```

Comment:
There is a Python built-in `filter()` so we should avoid redefining it just as a best practices thing. Additionally, and this behavior was not something I learned until recently, we should avoid defining data structures as default values as they are reused for all invocations of the function. In this case it does not matter but if we were to make changes to the default list that change would propagate across invocations.

```suggestion
    ec2: boto3.client, filters: Optional[List[Dict[str, Union[str, List[str]]]]] = None
) -> List[str]:
    """Get a filtered list of all the regions with EC2 support."""
    if filters is None:
        filters = []

    response = ec2.describe_regions(Filters=filters)
```